### PR TITLE
Support App environment Variables

### LINKED ISSUE #375: Set environment variables
Depends on #347 

Various applications may need to be configured through environment variables. For example, a rails application needs to have the `RAILS_MASTER_KEY` environment variable set to a specific secret (https://edgeguides.rubyonrails.org/security.html#custom-credentials).

## Acceptance Criteria

The user can specify env vars for a created app (i.e. when creating the app). See https://github.com/epinio/epinio/issues/375#issuecomment-868301329

When pushing the app, staging will create a deployment. The deployment will mount the secret as env vars.

mm: I'd suggest to postpone `--staging` env vars for a separate follow-up story.

## Previous Notes

Also buildpacks are moving away from `buildpack.yaml` configuration towards env vars (https://github.com/paketo-buildpacks/nginx/blob/main/rfcs/0001-buildpack-yml-to-env-var.md).

We need a way to allow the user to set environment variables that will be available to the application at run time.

We should also investigate if the same solution would work for staging with buildpacks (see the link above).

## Previous Questions

- Where are the variables stored? Is it on the deployment itself?
- What happens when we push the application again?
- How does the user delete a variable?
- Can the solution look similar to the concept of services? E.g. each environment variable being a secret behind the scenes that can be mounted to any application, kinda like a service, so it has its own lifecycle.

### LINKED ISSUE #510: Listing apps, should show application resources instead of deployments
Now listing apps, we see all the deployments. With the introduction of an application resource, that might exist before the deployment (workload), exists, we want the user to see it in the CLI output.

**Note**: The same might be true for other commands, like `app show`. We might need to create follow up stories.

relies on #509 

## Acceptance Criteria

* epinio show app includes apps without deployments

### PR DESCRIPTION
**Pull request by andreas-kupries**
Fix #375 
Fix #510

### Todo

  - More tests

### CODE REVIEW

**Review by thardeck — DISMISSED** [review]

File: internal/api/v1/applications.go [hunk 1]
```diff
@@ -222,24 +249,41 @@ func (hc ApplicationsController) Logs(w http.ResponseWriter, r *http.Request) {
...
 		if app == nil {
-			jsonErrorResponse(w, AppIsNotKnown(appName))
+			// While app exists it has no workload
+			jsonErrorResponse(w, NewAPIError("No logs available for application without workload", "", http.StatusBadRequest))
+			return
 		}
 	}
 
 	if appName == "" && stageID == "" {
 		jsonErrorResponse(w, BadRequest(errors.New("You need to speficy either the stage id or the app")))
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply internal/api/v1/applications.go:285]: will do

Comment (andreas-kupries) [inline internal/api/v1/applications.go:285]:
```suggestion
		jsonErrorResponse(w, BadRequest(errors.New("You need to specify either the stage id or the app")))
```

[the remark this answers is not in the document]
Comment (andreas-kupries) [reply internal/api/v1/applications.go]: And done, via github suggestion

**Review by thardeck — DISMISSED** [review]

File: internal/cli/env.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+		}
+
+		return nil
+	},
+	ValidArgsFunction: func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		if len(args) != 0 {
+			return nil, cobra.ShellCompDirectiveNoFileComp
+		}
+
+		// TODO: arg 1: EV match
```
Comment (manno) [inline internal/cli/env.go:177]: I don't understand this TODO

File: internal/cli/clients/client.go [hunk 3]
```diff
@@ -177,6 +319,10 @@ func (c *EpinioClient) ServicePlanMatching(ctx context.Context, serviceClassName
 
 	result := []string{}
 
+	// TODO Create and use server endpoints. Maybe use existing
```
Comment (manno) [inline internal/cli/clients/client.go:322]: Ah nice, these TODOs are for
https://github.com/epinio/epinio/issues/508

File: internal/cli/clients/client.go [hunk 4]
```diff
@@ -90,6 +90,148 @@ func NewEpinioClient(ctx context.Context, flags *pflag.FlagSet) (*EpinioClient,
 	return epinioClient, nil
 }
 
+// EnvList displays a table of all environment variables and their
+// values for the named application.
+func (c *EpinioClient) EnvList(ctx context.Context, appName string) error {
+	log := c.Log.WithName("EnvList")
+	log.Info("start")
+	defer log.Info("return")
+	// details := log.V(1) // NOTE: Increment of level, not absolute.
```
Comment (manno) [inline internal/cli/clients/client.go:99]:
```suggestion
```

File: internal/application/workload.go [hunk 5]
```diff
@@ -72,6 +134,10 @@ func (a *Workload) Scale(ctx context.Context, instances int32) error {
 
 // UnbindAll dissolves all bindings from the application.
 func (a *Workload) UnbindAll(ctx context.Context, cluster *kubernetes.Cluster, svcs []string) error {
+
+	// TODO: Optimize this action to perform a single restart for
```
Comment (manno) [inline internal/application/workload.go:138]: Maybe add an issue for this TODO and link it here. Though, 'optimize' issues don't make it in the backlog often...

File: internal/cli/env.go (hunk 2 above)
Comment (andreas-kupries) [reply internal/cli/env.go:177]: The command line completion should should complete the second argument (index 1) based on the existing environment variables of the app.

File: internal/cli/clients/client.go (hunk 3 above)
Comment (andreas-kupries) [reply internal/cli/clients/client.go:322]: Yes

File: internal/application/workload.go (hunk 5 above)
Comment (andreas-kupries) [reply internal/application/workload.go:138]: Done, #598

**Pull request merged**